crates/db_connection_pool/src/dbconnection/odbcconn.rs

```diff
@@ -0,0 +1,231 @@
+pub type ODBCDbConnectionPool<'a> =
+    dyn DbConnectionPool<Connection<'a>, ODBCParameter> + Sync + Send;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    ArrowODBCError { source: arrow_odbc::Error },
+    ODBCAPIError { source: odbc_api::Error },
+}
+
+pub struct ODBCConnection<'a> {
+    pub conn: Arc<Mutex<Connection<'a>>>,
+}
+
+impl<'a> DbConnection<Connection<'a>, ODBCParameter> for ODBCConnection<'a>
+where
+    'a: 'static,
+{
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn as_async(&self) -> Option<&dyn super::AsyncDbConnection<Connection<'a>, ODBCParameter>> {
+        Some(self)
+    }
+}
+
+#[async_trait::async_trait]
+impl<'a> AsyncDbConnection<Connection<'a>, ODBCParameter> for ODBCConnection<'a>
+where
+    'a: 'static,
+{
+    fn new(conn: Connection<'a>) -> Self {
+        ODBCConnection {
+            conn: Arc::new(conn.into()),
+        }
+    }
+
+    #[must_use]
+    async fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef> {
+        let cxn = self.conn.lock().await;
+
+        let mut prepared = cxn.prepare(&format!(
+            "SELECT * FROM {} LIMIT 1",
+            table_reference.to_quoted_string()
+        ))?;
+        let schema = Arc::new(arrow_schema_from(&mut prepared)?);
+
+        Ok(schema)
+    }
+
+    async fn query_arrow(
+        &self,
+        sql: &str,
+        params: &[ODBCParameter],
+    ) -> Result<SendableRecordBatchStream> {
+        let cxn = self.conn.lock().await;
+        let mut prepared = cxn.prepare(sql)?;
+        let schema = Arc::new(arrow_schema_from(&mut prepared)?);
+        let mut statement = prepared.into_statement();
+
+        bind_parameters(&mut statement, params);
+
+        let cursor = unsafe {
+            statement.execute().unwrap();
+            CursorImpl::new(statement.as_stmt_ref())
+        };
+
+        let reader = OdbcReaderBuilder::new()
+            .with_schema(schema.clone())
+            .build(cursor)
+            .context(ArrowODBCSnafu)?;
+
+        let results: Vec<RecordBatch> = reader.map(|b| b.unwrap()).collect();
+
+        Ok(Box::pin(MemoryStream::try_new(results, schema, None)?))
+    }
+
+    async fn execute(&self, query: &str, params: &[ODBCParameter]) -> Result<u64> {
+        let cxn = self.conn.lock().await;
+        let prepared = cxn.prepare(query)?;
+        let mut statement = prepared.into_statement();
+
+        bind_parameters(&mut statement, params);
+
+        let row_count = unsafe {
+            statement.execute().unwrap();
+            statement.row_count()
+        };
+
+        Ok(row_count
+            .unwrap()
+            .try_into()
+            .expect("Must obtain row count"))
+    }
+}
+
+fn bind_parameters(statement: &mut StatementImpl, params: &[ODBCParameter]) {
+    for (i, param) in params.iter().enumerate() {
+        unsafe {
```

Comment:
Same here. Why is `unsafe` necessary here?

Reply:
[odbc-api reference for parameters](https://docs.rs/odbc-api/7.0.0/odbc_api/parameter/index.html)

[`bind_input_parameter`](https://docs.rs/odbc-api/7.0.0/odbc_api/handles/trait.Statement.html#method.bind_input_parameter) is unsafe. The crate has a safe API for doing this (like we are by exposing a safe function that leverages unsafe code) but I cannot use it given the constraints of `AsyncDbConnection` and also `async_trait`.

```rust
pub trait AsyncDbConnection<T, P>: DbConnection<T, P> + Sync {
    ...
    async fn query_arrow(&self, sql: &str, params: &[P]) -> Result<SendableRecordBatchStream>;
    async fn execute(&self, sql: &str, params: &[P]) -> Result<u64>;
}
```

Our input parameters need to be sync to work with `async_trait`:
```rust
// initially: pub type ODBCParameter = Box<dyn InputParameter + Sync>
pub trait ODBCSyncParameter: InputParameter + Sync {...}
```

But this doesn't work for [`execute`](https://docs.rs/odbc-api/7.0.0/odbc_api/struct.Connection.html#method.execute) (the 'safe' API), and I have no way to rebox the ref to cast to `&dyn InputParameter`:

```
the trait bound `[Box<dyn InputParameter + Sync>]: odbc_api::parameter_collection::InputParameterCollection` is not satisfied
the trait `odbc_api::parameter_collection::InputParameterCollection` is implemented for `[T]`
required for `&[Box<dyn InputParameter + Sync>]` to implement `ParameterCollectionRef`
```

The odbc-api [IntoParameter](https://docs.rs/odbc-api/7.0.0/odbc_api/trait.IntoParameter.html) trait wants to consume an instance to lift into `InputParameter` (common pattern for FFI glue for data going out). I think it would be worthwhile to think about passing params by value. It would let us use the safe API directly:

```
async fn query_arrow(&self, sql: &str, params: [P]) -> Result<SendableRecordBatchStream>;
async fn execute(&self, sql: &str, params: [P]) -> Result<u64>;
```

I am happy to work with you to figure out a cleaner way around this, I spent a bunch of time here but unfortunately I wasn't able to find a better solution. Let me know if this makes sense!

Reply:
Thanks for detailed feedback. Makes sense. 
I think we only pass empty vec params to `execute` and `query_arrow` as of now.
Probably can make a change to pass params by value, but needs some more investigating. I will look into it.
Also, would be nice to see comments with justifications of using unsafe in a code.